

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Metagenome Prediction Tutorial &mdash; PICRUSt 0.9.1-dev documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.9.1-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PICRUSt 0.9.1-dev documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">PICRUSt 0.9.1-dev documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="metagenome-prediction-tutorial">
<span id="id1"></span><h1>Metagenome Prediction Tutorial<a class="headerlink" href="#metagenome-prediction-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This tutorial explains how to predict a microbial community metagenome using PICRUST, based on 16S (or other marker gene) data as detailed in <a class="reference internal" href="otu_picking.html#otu-picking-tutorial"><em>Picking OTUs for use in PICRUSt</em></a>.</p>
<p><strong>BIOM Format</strong></p>
<ul class="simple">
<li>Please note that PICRUSt by default uses the relatively new <a class="reference external" href="http://biom-format.org">biom</a> format for representing OTU tables and Gene tables (e.g. KOs by samples). This <a class="reference external" href="http://biom-format.org/documentation/biom_format.html#motivation-for-the-biom-format">has several benefits</a> including easier integration with other software (e.g. QIIME and others in the future) and allows embedding of extra metadata about both the samples and observations (OTUs/KOs).</li>
<li><strong>However, PICRUSt also allow users to input OTU tables and export PICRUSt predictions in tab-delimited format by using the &#8216;-f&#8217; option (see below).</strong></li>
<li>In addition, users can always convert to/from biom format to tab-delimited format using BIOM&#8217;s built in <a class="reference external" href="http://biom-format.org/documentation/biom_conversion.html">conversion script</a>.</li>
</ul>
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>You should have already installed PICRUSt (<a class="reference internal" href="../install.html#install"><em>Installing PICRUSt</em></a>).</li>
<li>A PICRUSt compatible OTU table (<a class="reference internal" href="otu_picking.html#otu-picking-tutorial"><em>Picking OTUs for use in PICRUSt</em></a>), such as the example file <strong>tutorials/hmp_mock_16S.biom</strong>.</li>
</ol>
</div>
<div class="section" id="step-1-normalize-otu-table">
<h2>Step 1: Normalize OTU Table<a class="headerlink" href="#step-1-normalize-otu-table" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="../scripts/normalize_by_copy_number.html#normalize-by-copy-number"><em>normalize_by_copy_number.py</em></a> normalizes the OTU table by dividing each OTU by the known/predicted 16S copy number abdundance.</p>
<p>Input is the users OTU table (that has been referenced picked against Greengenes).</p>
<p>Input and output files are in <a class="reference external" href="http://biom-format.org">biom</a> format:</p>
<div class="highlight-python"><pre>normalize_by_copy_number.py
        -i hmp_mock_16S.biom
        -o normalized_otus.biom</pre>
</div>
<p><strong>(Optional) Input format of OTU table can be changed to tab-delimited &#8220;classic&#8221; OTU table instead of BIOM format using the &#8216;-f&#8217; option:</strong></p>
<div class="highlight-python"><pre>normalize_by_copy_number.py
       -f
       -i hmp_mock_16S.tab
       -o normalized_otus.biom</pre>
</div>
</div>
<div class="section" id="step-2-predict-functions-for-metagenome">
<h2>Step 2: Predict Functions For Metagenome<a class="headerlink" href="#step-2-predict-functions-for-metagenome" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="../scripts/predict_metagenomes.html#predict-metagenomes"><em>predict_metagenomes.py</em></a> creates the final metagenome functional predictions. It multiplies each normalized OTU abundance by each predicted functional trait abundance to produce a table of functions (rows) by samples (columns). (Note: This step requires 2-3GB of RAM)</p>
<p>Input is the normalized OTU table created by <a class="reference internal" href="../scripts/normalize_by_copy_number.html#normalize-by-copy-number"><em>normalize_by_copy_number.py</em></a>.</p>
<p>Output is in <a class="reference external" href="http://biom-format.org">biom</a> format by default:</p>
<div class="highlight-python"><pre>predict_metagenomes.py
        -i normalized_otus.biom
        -o metagenome_predictions.biom</pre>
</div>
<p><strong>(Optional) Output format can be changed from BIOM to tab delimited using &#8216;-f&#8217; option:</strong></p>
<div class="highlight-python"><pre>predict_metagenomes.py
        -f
        -i normalized_otus.biom
        -o metagenome_predictions.tab</pre>
</div>
<p>(Optional) NSTI values for each sample can be obtained using the <tt class="docutils literal"><span class="pre">-a</span></tt> option:</p>
<div class="highlight-python"><pre>predict_metagenomes.py
        -i normalized_otus.biom
        -o metagenome_predictions.tab
        -a nsti_per_sample.tab</pre>
</div>
<p>(Optional) Those with machines &gt; 6GB of RAM can get faster run times (~50% less) using the <tt class="docutils literal"><span class="pre">--suppress_subset_loading</span></tt> option:</p>
<div class="highlight-python"><pre>predict_metagenomes.py
        --suppress_subset_loading
        -i normalized_otus.biom
        -o metagenome_predictions.biom</pre>
</div>
</div>
<div class="section" id="step-3-analysing-predicted-metagenomes">
<h2>Step 3: Analysing Predicted Metagenomes<a class="headerlink" href="#step-3-analysing-predicted-metagenomes" title="Permalink to this headline">¶</a></h2>
<p>Know that you have your predicted metagenome you can analyse it just the same as a real metagenome.
Some suggestions for downstream analysis are outlined in <a class="reference internal" href="downstream_analysis.html#downstream-analysis-guide"><em>Analyzing PICRUSt predicted metagenomes</em></a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Metagenome Prediction Tutorial</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#step-1-normalize-otu-table">Step 1: Normalize OTU Table</a></li>
<li><a class="reference internal" href="#step-2-predict-functions-for-metagenome">Step 2: Predict Functions For Metagenome</a></li>
<li><a class="reference internal" href="#step-3-analysing-predicted-metagenomes">Step 3: Analysing Predicted Metagenomes</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tutorials/metagenome_prediction.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">PICRUSt 0.9.1-dev documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011-2013, The PICRUSt Project.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2pre.
    </div>
  </body>
</html>